p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,281 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"crypto"
+	"encoding/hex"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/udpmux"
+
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multibase"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/pion/ice/v2"
+	"github.com/pion/webrtc/v3"
+)
+
+var _ tpt.Listener = &listener{}
+
+const (
+	maxConnectionBufferSize int = 10
+	candidateSetupTimeout       = 20 * time.Second
+	candidateChanSize           = 20
+)
+
+type candidateAddr struct {
+	ufrag string
+	raddr *net.UDPAddr
+}
+
+type listener struct {
+	transport                 *WebRTCTransport
+	config                    webrtc.Configuration
+	localFingerprint          webrtc.DTLSFingerprint
+	localFingerprintMultibase string
+	mux                       ice.UDPMux
+	ctx                       context.Context
+	cancel                    context.CancelFunc
+	localAddr                 net.Addr
+	localMultiaddr            ma.Multiaddr
+	connChan                  chan tpt.CapableConn
+	wg                        sync.WaitGroup
```

Comment:
Group `wg`, `ctx`, and `cancel`, and a comment that they control the lifecycle of this listener.